File: Android/app/src/main/java/com/khushi/conduit/link/LinkConnection.kt
package com.khushi.conduit.link

import android.util.Log
import com.khushi.conduit.core.link.LinkSigner
import com.khushi.conduit.core.link.PhoneHandshake
import com.khushi.conduit.core.protocol.ControlType
import com.khushi.conduit.core.protocol.DeviceInfo
import com.khushi.conduit.core.protocol.Envelope
import com.khushi.conduit.core.protocol.ErrorCode
import com.khushi.conduit.core.protocol.Frame
import com.khushi.conduit.core.protocol.FrameDecoder
import com.khushi.conduit.core.protocol.HandshakeIntent
import com.khushi.conduit.core.protocol.HandshakeMessage
import com.khushi.conduit.core.protocol.LinkChannel
import com.khushi.conduit.core.protocol.LinkCrypto
import com.khushi.conduit.core.protocol.LinkSessionInfo
import com.khushi.conduit.core.protocol.ProtocolError
import java.io.IOException
import java.io.OutputStream
import java.net.InetSocketAddress
import java.net.Socket

/**
 * One connection from this phone to a Mac: the socket, the frames, the
 * handshake driven to ready, and sealed envelopes after it.
 *
 * [run] blocks its thread until the connection ends, so callers give it one.
 * The two hellos are plaintext; everything after is sealed, and a frame that
 * does not open ends the connection rather than being guessed at.
 */
class LinkConnection(
    signer: LinkSigner,
    device: DeviceInfo,
    intent: HandshakeIntent,
    expectedMacKey: ByteArray?,
    private val listener: Listener,
) {

    interface Listener {
        fun onPairingCode(code: String, mac: DeviceInfo) {}
        fun onReady(connection: LinkConnection, mac: DeviceInfo, macKey: ByteArray, session: LinkSessionInfo) {}
        fun onEnvelope(connection: LinkConnection, envelope: Envelope) {}
        fun onClosed(error: ProtocolError?) {}
    }

    private val handshake = PhoneHandshake(signer, device, intent, expectedMacKey)
    private val socket = Socket()
    private val decoder = FrameDecoder()
    private var output: OutputStream? = null
    private var sealer: LinkCrypto.Sealer? = null
    private var opener: LinkCrypto.Opener? = null
    @Volatile private var ready = false
    @Volatile private var closed = false
    private var closeError: ProtocolError? = null

    /** Connect and serve until the connection ends. */
    fun run(host: String, port: Int, connectTimeoutMs: Int = 4_000) {
        try {
            socket.connect(InetSocketAddress(host, port), connectTimeoutMs)
            socket.tcpNoDelay = true
            // The handshake has a minute; after it, silence longer than three
            // missed heartbeats means the Mac has gone.
            socket.soTimeout = 60_000
            output = socket.getOutputStream()
            perform(handshake.start())

            val input = socket.getInputStream()
            val buffer = ByteArray(64 * 1024)
            while (!closed) {
                val count = input.read(buffer)
                if (count < 0) break
                for (frame in decoder.receive(buffer.copyOf(count))) {
                    if (closed) break
                    handle(frame)
                }
            }
        } catch (error: IOException) {
            if (!closed) closeError = ProtocolError(ErrorCode.DEVICE_UNAVAILABLE, "The Mac could not be reached.")
            Log.i(TAG, "connection ended: ${error.javaClass.simpleName}")
        } catch (error: Exception) {
            if (!closed) closeError = ProtocolError(ErrorCode.INVALID_REQUEST, "The Mac sent something Conduit could not read.")
            Log.w(TAG, "connection failed", error)
        } finally {
            finish()
        }
    }

    fun confirmPairing() = perform(handshake.confirmPairing())

    fun rejectPairing() = perform(handshake.rejectPairing())

    fun send(envelope: Envelope) {
        if (!ready) return
        send(envelope.encode().encodeToByteArray(), ControlType.ENVELOPE, sealed = true)
    }

    /** Heartbeat interval agreed at ready; reads give up after three missed. */
    fun useHeartbeat(seconds: Int) {
        runCatching { socket.soTimeout = seconds * 3_000 }
    }

    fun close(error: ProtocolError? = null) {
        if (closed) return
        closeError = closeError ?: error
        closed = true
        runCatching { socket.close() }
    }

    private fun handle(frame: Frame) {
        val payload = opener?.let {
            try {
                it.open(frame.payload, frame.channel.value, frame.type)
            } catch (_: Exception) {
                close(ProtocolError(ErrorCode.INTERNAL, "A message from the Mac could not be opened."))
                return
            }
        } ?: frame.payload

        when {
            frame.channel == LinkChannel.CONTROL && frame.type == ControlType.HANDSHAKE -> {
                if (ready) return close(ProtocolError(ErrorCode.INVALID_REQUEST, "The Mac tried to hand shake twice."))
                val message = runCatching { HandshakeMessage.decode(payload) }.getOrNull()
                    ?: return close(ProtocolError(ErrorCode.INVALID_REQUEST, "The Mac sent an unreadable handshake."))
                perform(handshake.receive(message, payload))
            }
            frame.channel == LinkChannel.CONTROL && frame.type == ControlType.ENVELOPE -> {
                if (!ready) return close(ProtocolError(ErrorCode.NOT_PAIRED, "The Mac sent messages before the handshake finished."))
                val envelope = runCatching { Envelope.decode(payload.decodeToString()) }.getOrNull()
                    ?: return close(ProtocolError(ErrorCode.INVALID_REQUEST, "The Mac sent an unreadable message."))
                listener.onEnvelope(this, envelope)
            }
            // Other channels arrive with their features.
            else -> Unit
        }
    }

    private fun perform(actions: List<PhoneHandshake.Action>) {
        for (action in actions) {
            when (action) {
                is PhoneHandshake.Action.Send -> send(action.payload, ControlType.HANDSHAKE, action.sealed)
                is PhoneHandshake.Action.KeysEstablished -> {
                    // The phone seals with its own key and opens with the Mac's.
                    sealer = LinkCrypto.Sealer(action.keys.phoneToMac)
                    opener = LinkCrypto.Opener(action.keys.macToPhone)
                }
                is PhoneHandshake.Action.ConfirmPairing -> listener.onPairingCode(action.code, action.mac)
                is PhoneHandshake.Action.Ready -> {
                    ready = true
                    listener.onReady(this, action.mac, action.macIdentityKey, action.session)
                }
                is PhoneHandshake.Action.Fail -> close(action.error)
            }
        }
    }

    private fun send(payload: ByteArray, type: Int, sealed: Boolean) {
        if (closed) return
        val body = if (sealed) {
            sealer?.seal(payload, LinkChannel.CONTROL.value, type)
                ?: return close(ProtocolError(ErrorCode.INTERNAL, "Conduit tried to seal a message before the keys existed."))
        } else {
            payload
        }
        try {
            synchronized(socket) { output?.run { write(Frame(LinkChannel.CONTROL, type, body).encode()); flush() } }
        } catch (_: IOException) {
            close(ProtocolError(ErrorCode.DEVICE_UNAVAILABLE, "The Mac stopped answering."))
        }
    }

    private fun finish() {
        closed = true
        runCatching { socket.close() }
        listener.onClosed(closeError)
    }

    private companion object {
        const val TAG = "ConduitLink"
    }
}

File: Shared/ConduitKit/Sources/ConduitCore/Link/LinkConnection.swift
//
//  LinkConnection.swift
//  ConduitCore
//
//  One phone's connection to this Mac: frames in, frames out, the handshake
//  driven to ready, and sealed envelopes after it.
//
//  The two hellos are plaintext; everything after them is sealed under the
//  session keys, so a frame that fails to open closes the connection rather
//  than being guessed at. The phone dials the Mac, because the phone is the
//  one that roams between networks.
//

import ConduitMedia
import ConduitProtocol
import Foundation
import Network

@MainActor
final class LinkConnection {

    /// Asks the person to compare the six digits before a new phone is trusted.
    struct PairingRequest {
        let code: String
        let device: DeviceInfo
    }

    var onPairing: (PairingRequest) -> Void = { _ in }
    var onReady: (LinkPeer) -> Void = { _ in }
    var onEnvelope: (Envelope, LinkPeer) -> Void = { _, _ in }
    var onClosed: (LinkPeer?, ProtocolError?) -> Void = { _, _ in }

    /// The phone on the other end, once the handshake has proved it.
    var peerID: String? { peer?.id }

    private let connection: NWConnection
    private let handshake: LinkHandshake
    private var decoder = FrameDecoder()
    private var sealer: LinkCrypto.Sealer?
    private var opener: LinkCrypto.Opener?
    private var peer: LinkPeer?
    private var closed = false
    /// Set once a refusal is on its way out: nothing more is read or handled.
    private var closing = false

    init(connection: NWConnection, identity: LinkIdentity, device: DeviceInfo,
         trust: LinkTrust, pairingOpen: Bool) {
        self.connection = connection
        handshake = LinkHandshake(identity: identity, device: device, trust: trust, pairingOpen: pairingOpen)
    }

    func start() {
        connection.stateUpdateHandler = { [weak self] state in
            Task { @MainActor [weak self] in
                guard let self else { return }
                switch state {
                case .failed(let error):
                    finish(ProtocolError(.internal, error.localizedDescription))
                case .cancelled:
                    finish(nil)
                default:
                    break
                }
            }
        }
        connection.start(queue: .main)
        receive()
    }

    /// The person compared the code and accepted, or refused.
    func confirmPairing() { perform(handshake.confirmPairing()) }
    func rejectPairing() { perform(handshake.rejectPairing()) }

    func send(_ envelope: Envelope) {
        guard let payload = try? envelope.encoded() else { return }
        send(payload, channel: .control, type: ControlType.envelope.rawValue, sealed: true)
    }

    func close(_ error: ProtocolError? = nil) {
        finish(error)
    }

    // MARK: - Receiving

    private func receive() {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 64 * 1024) { [weak self] data, _, isComplete, error in
            Task { @MainActor [weak self] in
                guard let self, !closed, !closing else { return }
                if let data, !data.isEmpty { decode(data) }
                if isComplete || error != nil {
                    finish(error.map { ProtocolError(.internal, $0.localizedDescription) })
                } else {
                    receive()
                }
            }
        }
    }

    private func decode(_ data: Data) {
        let frames: [Frame]
        do {
            frames = try decoder.receive(data)
        } catch {
            // TCP does not reorder or drop bytes, so this is a bug or an
            // impostor. Either way the stream cannot be trusted any more.
            finish(ProtocolError(.invalidRequest, "The phone sent a frame Conduit could not read."))
            return
        }
        for frame in frames where !closed && !closing {
            handle(frame)
        }
    }

    private func handle(_ frame: Frame) {
        guard let payload = plaintext(of: frame) else { return }

        switch (frame.channel, frame.type) {
        case (.control, ControlType.handshake.rawValue):
            guard peer == nil else {
                finish(ProtocolError(.invalidRequest, "The phone tried to hand shake twice."))
                return
            }
            guard let message = try? HandshakeMessage.decode(payload) else {
                finish(ProtocolError(.invalidRequest, "The phone sent an unreadable handshake."))
                return
            }
            perform(handshake.receive(message, raw: payload))

        case (.control, ControlType.envelope.rawValue):
            guard let peer else {
                finish(ProtocolError(.notPaired, "The phone sent messages before the handshake finished."))
                return
            }
            guard let envelope = try? Envelope.decode(payload) else {
                finish(ProtocolError(.invalidRequest, "The phone sent an unreadable message."))
                return
            }
            onEnvelope(envelope, peer)

        default:
            // Channels for input, haptics and files arrive with their
            // features; until then, silence is better than a guess.
            CoreLog.engine.debug("ignoring a Conduit Link frame on channel \(frame.channel.rawValue)")
        }
    }

    /// Frames are plaintext until the session keys exist, and sealed after.
    private func plaintext(of frame: Frame) -> Data? {
        guard opener != nil else { return frame.payload }
        do {
            return try opener?.open(frame.payload, channel: frame.channel.rawValue, type: frame.type)
        } catch {
            finish(ProtocolError(.internal, "A message from the phone could not be opened."))
            return nil
        }
    }

    // MARK: - Sending

    private func perform(_ actions: [LinkHandshake.Action]) {
        for action in actions {
            switch action {
            case .send(_, let payload, let sealed):
                send(payload, channel: .control, type: ControlType.handshake.rawValue, sealed: sealed)
            case .confirmPairing(let code, let device):
                onPairing(PairingRequest(code: code, device: device))
            case .keysEstablished(let keys):
                // From here on the Mac seals with its own key and opens with
                // the phone's.
                sealer = LinkCrypto.Sealer(key: keys.macToPhone)
                opener = LinkCrypto.Opener(key: keys.phoneToMac)
            case .ready(let peer):
                self.peer = peer
                onReady(peer)
            case .fail(let error):
                finishAfterSending(error)
            }
        }
    }

    /// Close once everything queued has gone out. MEASURED: cancelling an
    /// NWConnection drops what it has not yet sent, so a phone that was
    /// refused saw the connection vanish without the error that says why.
    private func finishAfterSending(_ error: ProtocolError?) {
        guard !closed, !closing else { return }
        closing = true
        connection.send(content: nil, contentContext: .finalMessage, isComplete: true,
                        completion: .contentProcessed { [weak self] _ in
                            Task { @MainActor [weak self] in self?.finish(error) }
                        })
    }

    private func send(_ payload: Data, channel: LinkChannel, type: UInt8, sealed: Bool) {
        guard !closed else { return }
        var body = payload
        if sealed {
            do {
                guard let sealed = try sealer?.seal(payload, channel: channel.rawValue, type: type) else {
                    finish(ProtocolError(.internal, "Conduit tried to seal a message before the keys existed."))
                    return
                }
                body = sealed
            } catch {
                finish(ProtocolError(.internal, "Conduit could not seal a message."))
                return
            }
        }
        connection.send(content: Frame(channel: channel, type: type, payload: body).encoded(),
                        completion: .contentProcessed { _ in })
    }

    private func finish(_ error: ProtocolError?) {
        guard !closed else { return }
        closed = true
        connection.cancel()
        onClosed(peer, error)
    }
}
